bentoml/deployment/serverless/__init__.py

```diff
@@ -0,0 +1,232 @@
+from bentoml.utils.whichcraft import which
+from bentoml.utils.exceptions import BentoMLException
+from bentoml.deployment.serverless.aws_lambda_template import create_aws_lambda_bundle, \
+    DEFAULT_AWS_DEPLOY_STAGE, DEFAULT_AWS_REGION
+from bentoml.deployment.serverless.gcp_function_template import create_gcp_function_bundle, \
+    DEFAULT_GCP_REGION, DEFAULT_GCP_DEPLOY_STAGE
+from bentoml.deployment.utils import generate_bentoml_deployment_snapshot_path
+
+
+logger = logging.getLogger(__name__)
+
+SERVERLESS_PROVIDER = {
+    'aws-lambda': 'aws-python3',
+    'aws-lambda-py2': 'aws-python',
+    'gcp-function': 'google-python',
+}
+
+DEFAULT_GCP_REGION = 'us-west2'
+
+
+def check_serverless_compatiable_version():
+    if which('serverless') is None:
+        raise ValueError('Serverless framework is not installed, please visit ' +
+                         'www.serverless.com for install instructions.')
+
+    version_result = subprocess.check_output(['serverless', '-v'])
+    parsed_version = version.parse(version_result.decode('utf-8').strip())
+
+    if parsed_version >= version.parse('1.40.0'):
+        return
+    else:
+        raise ValueError(
+            'Incompatiable serverless version, please install version 1.40.0 or greater')
+
+
+def parse_serverless_response(serverless_response):
+    """Parse serverless response string, raise error if it is a serverless error,
+    otherwise, return information.
+    """
+    str_list = serverless_response.strip().split('\n')
+    error = [s for s in str_list if 'Serverless Error' in s]
+    if error:
+        error_pos = str_list.index(error[0])
+        error_message = str_list[error_pos + 1]
+        raise BentoMLException(error_message)
+    return str_list
+
+class ServerlessDeployment(object):
+    """Managing deployment operations for serverless
+    """
+
+    def __init__(self, platform, archive_path, region, stage):
+        check_serverless_compatiable_version()
+
+        self.platform = platform
+        self.provider = SERVERLESS_PROVIDER[platform]
+        self.archive_path = archive_path
+        self.bento_service = load(archive_path)
+        if platform == 'google-python':
+            self.region = DEFAULT_GCP_REGION if region is None else region
+            self.stage = DEFAULT_GCP_DEPLOY_STAGE if stage is None else stage
+        elif platform == 'aws-lambda' or platform == 'aws-lambda-py':
+            self.region = DEFAULT_AWS_REGION if region is None else region
+            self.stage = DEFAULT_AWS_DEPLOY_STAGE if stage is None else stage
+
+    def generate_bundle(self):
+        output_path = generate_bentoml_deployment_snapshot_path(self.bento_service.name, self.platform)
+        Path(output_path).mkdir(parents=True, exist_ok=False)
+
+        # Calling serverless command to generate templated project
+        with subprocess.Popen(['serverless', 'create', '--template', self.provider, '--name', self.bento_service.name],
```

Comment:
just curious what's the difference between this and 
```
p = subprocess.Popen(...)
stdout, stderr = p.communicate
```
which seems more common?

Reply:
For our current use cases, we don't have a lot of interactions with the process, this style is fine.  We don't have to terminate the process ourselves. 

For more advanced use cases, we probably want to utilize the style in your comment's code block, since we can have more control over input/output like add more input into the process (username/password/etc).